rfcs/002 - Side-effects in Marsha.md

```diff
@@ -0,0 +1,126 @@
+```
+
+This blends input and output side effects together and makes it a more flexible notice the user provides. It is not much different from the existing `cnn.mrsh` *except* hopefully you don't have to explicitly call out what has changed. It would depend on extra LLM calls to determine what the side-effect sources and destinations are, and then do the automated work to get an example of each source to use.
+
+#### Disallow mocks in the test suite
+
+One possibility to fix this situation is to disable mocking in the generated test suite at all. In that case, the side-effect functions *will* be executed for real. This could resolve the side-effect input problem by using real-world data for the test, but it would make the reliability of the generation lower (why we added mock support in the first place), It also doesn't do anything for the side-effect outputs, which would remain completely untested.
+
+#### Explicit mock rule declaration in examples
+
+Only generate mocks for things that were called out to be mocked and/or intercepted for side-effects inputs and outputs, respectively. This would similarly need extra work to get the actual payload to embed in the mocks.
+
+```md
+# func cnn(string of section to take headlines from): list of headlines
+
+This function scrapes the cnn.com website for headlines. The section it takes the headlines from is passed to it, with 'home' referring to the homepage at cnn.com and should be special cased, while 'us' refers to cnn.com/us, 'politics' refers to cnn.com/politics, and so on for every top-level category CNN has.
+
+* cnn('home') using a mock of https://cnn.com = ["Florida's new standards for teaching Black history spark outrage", "His books sold over 300 million copies and were translated into 63 languages. Now, a museum is acknowledging his racism", "Player quits match in tears as tennis world slams opponent’s ‘absolutely disgusting’ actions"]
+* cnn('us') using a mock of https://cnn.com/us = ['18-year-old Miami woman arrested after allegedly trying to hire a hitman to go after her 3-year-old son', 'Investigation into Gilgo Beach serial killings suspect expands to Nevada and South Carolina', 'Rescue crews continue search for 2 children swept away by Pennsylvania floodwater that killed their mother']
+* cnn('world') using a mock of https://cnn.com/world = ["Police raids follow shocking video of sexual assault in India’s Manipur state amid ethnic violence", 'Ukrainian air defenses in Odesa outgunned as Russia targets global grain supply', 'Anger boils over as Kenya’s cost of living protests shake the nation']
+```
+
+This approach seems pretty "natural" with the examples, bolting on some special behavior if you follow a particular pattern in the example set, but that may also reduce the discoverability of the mocking mechanism, or implying other features are automatically possible by using similar phrasing for other things.
```

Comment:
this option seems interesting. I agree that could reduce the discoverability but not sure if that should be a blocker. What I'm not so sure about is if it would be too verbose/repetitive if all the examples use the same mock. Now, if each example could use a different mock, then this one or a similar alternative seems to be the way to go.

Reply:
I think the hard part is the one related to side-effect outputs, I think none of the current alternatives is able to solve it?